### Student Performance Indicator

Student Performance Indicator

1.Life cycle of Machine learning Project

2.Understanding the Problem Statement

3.Data Collection

4.Data Checks to perform

5.Exploratory data analysis

6.Data Pre-Processing

7.Model Training

8.Choose best model


#1) Problem statement

This project understands how the student's performance (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.

2).Data Collection
Dataset Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977

The data consists of 8 column and 1000 rows.

In [5]:
#Import the packages
import pandas as pd
import numpy as np          
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [6]:
# load the dataset
df=pd.read_csv('./stud.csv')
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [7]:
df.shape


(1000, 8)

2.Dataset Information

1.gender : sex of students -> (Male/female)

2.race/ethnicity : ethnicity of students -> (Group A, B,C, D,E)

3.parental level of education : parents' final education ->(bachelor's degree,some college,master's degree,associate's degree,high school)

4.lunch : having lunch before test (standard or free/reduced)

5.test preparation course : complete or not complete before test

6.math score

7.reading score

8.writing score

# 3. Data Checks to perform
### Check Missing values
### Check Duplicates
### Check data type
### Check the number of unique values of each column
### Check statistics of data set
### Check various categories present in the different categorical column

### check missing values

In [10]:
# df.isnull().sum()
df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

### check for duplicates


In [11]:
df.duplicated().sum()

np.int64(0)

### check for data types


In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


### Checking the number of unique values of each column

In [18]:
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

### Check statistics of data set

In [19]:
df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

## Insight

#### From above description of numerical data, all means are very close to each other - between 66 and 68.05;
#### All standard deviations are also close - between 14.6 and 15.19;
#### While there is a minimum score 0 for math, for writing minimum is much higher = 10 and for reading myet higher = 17

# Exploring Data

In [20]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [21]:
print("Categories in 'geneder' variable:  ",end='')
print(df['gender'].unique())
print("Categories in 'race_ethnicity' variable:  ",end='')
print(df['race_ethnicity'].unique())
print("Categories in 'parental_level_of_education' variable:  ",end='')
print(df['parental_level_of_education'].unique())
print("Categories in 'lunch' variable: ",end='')
print(df['lunch'].unique())

Categories in 'geneder' variable:  ['female' 'male']
Categories in 'race_ethnicity' variable:  ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in 'parental_level_of_education' variable:  ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in 'lunch' variable: ['standard' 'free/reduced']


In [23]:
#define numerical and categorical columns
numeric_features=[feature for feature in df.columns if df[feature].dtype!='O']
categorical_features=[feature for feature in df.columns if df[feature].dtype=='O']

print(f"we have {len(numeric_features)} Numericalfeatures :{numeric_features}")
print(f"we have {len(categorical_features)} Categoricalfeatures :{categorical_features}")   

we have 3 Numericalfeatures :['math_score', 'reading_score', 'writing_score']
we have 5 Categoricalfeatures :['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']


In [24]:
df.head(2)

gender race_ethnicity parental_level_of_education     lunch  \
0  female        group B           bachelor's degree  standard   
1  female        group C                some college  standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88

## Adding columns for "Total Score" and "Average"

In [25]:
df['total_score']=df['math_score']+df['reading_score']+df['writing_score']
df['average_score']=df['total_score']/3

In [26]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total_score  average_score  
0          218      72.666667  
1          247      82.333333  
2          278      92.666667  
3          148      49.333333  
4          229      76.333333

In [29]:
reading_full=df[df['reading_score']==100]['average_score'].count()
writing_full=df[df['writing_score']==100]['average_score'].count()
maths_full=df[df['math_score']==100]['average_score'].count()

print(f'Number of students with full marks in Maths: {maths_full}')
print(f'Number of students with full marks in Writing: {writing_full}')
print(f'Number of students with full marks in Reading: {reading_full}')

Number of students with full marks in Maths: 7
Number of students with full marks in Writing: 14
Number of students with full marks in Reading: 17


In [31]:
reading_less=df[df['reading_score']<=20]['average_score'].count()
writing_less=df[df['writing_score']<=20]['average_score'].count()
maths_less=df[df['math_score']<=20]['average_score'].count()
print(f'Number of students with less than 20 marks in Maths: {maths_less}')
print(f'Number of students with less than 20 marks in Writing: {writing_less}')
print(f'Number of students with less than 20 marks in Reading: {reading_less}')

Number of students with less than 20 marks in Maths: 4
Number of students with less than 20 marks in Writing: 3
Number of students with less than 20 marks in Reading: 1


### Insights
### From above values we get students have performed the worst in Maths
### Best performance is in reading section

### 4. Exploring Data ( Visualization )
### 4.1 Visualize average score distribution to make some conclusion.
### Histogram
### Kernel Distribution Function (KDE)

In [ ]:
fig,axs=plt.subplot(1,2,figsize=(15,7))
plt.subplot(121)
sns.histplot(data=df,x='average',bins=30,kde=True,color='g')
plt.subplot(122)
sns.histplot(data=df,x='average',kde=True,hue='gender')
plt.show()

In [ ]:
print('helo')